third_part/openssl/ssl/man/man3/buffer.3
.\" Automatically generated by Pod::Man 2.22 (Pod::Simple 3.07)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "buffer 3"
.TH buffer 3 "2014-09-04" "1.0.1e" "OpenSSL"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
BUF_MEM_new, BUF_MEM_free, BUF_MEM_grow, BUF_strdup \- simple
character arrays structure
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 1
\& #include <openssl/buffer.h>
\&
\& BUF_MEM *BUF_MEM_new(void);
\&
\& void   BUF_MEM_free(BUF_MEM *a);
\&
\& int    BUF_MEM_grow(BUF_MEM *str, int len);
\&
\& char * BUF_strdup(const char *str);
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
The buffer library handles simple character arrays. Buffers are used for
various purposes in the library, most notably memory BIOs.
.PP
The library uses the \s-1BUF_MEM\s0 structure defined in buffer.h:
.PP
.Vb 6
\& typedef struct buf_mem_st
\& {
\&        int length;     /* current number of bytes */
\&        char *data;
\&        int max;        /* size of buffer */
\& } BUF_MEM;
.Ve
.PP
\&\fBlength\fR is the current size of the buffer in bytes, \fBmax\fR is the amount of
memory allocated to the buffer. There are three functions which handle these
and one \*(L"miscellaneous\*(R" function.
.PP
\&\fIBUF_MEM_new()\fR allocates a new buffer of zero size.
.PP
\&\fIBUF_MEM_free()\fR frees up an already existing buffer. The data is zeroed
before freeing up in case the buffer contains sensitive data.
.PP
\&\fIBUF_MEM_grow()\fR changes the size of an already existing buffer to
\&\fBlen\fR. Any data already in the buffer is preserved if it increases in
size.
.PP
\&\fIBUF_strdup()\fR copies a null terminated string into a block of allocated
memory and returns a pointer to the allocated block.
Unlike the standard C library \fIstrdup()\fR this function uses \fIOPENSSL_malloc()\fR and so
should be used in preference to the standard library \fIstrdup()\fR because it can
be used for memory leak checking or replacing the \fImalloc()\fR function.
.PP
The memory allocated from \fIBUF_strdup()\fR should be freed up using the \fIOPENSSL_free()\fR
function.
.SH "RETURN VALUES"
.IX Header "RETURN VALUES"
\&\fIBUF_MEM_new()\fR returns the buffer or \s-1NULL\s0 on error.
.PP
\&\fIBUF_MEM_free()\fR has no return value.
.PP
\&\fIBUF_MEM_grow()\fR returns zero on error or the new size (i.e. \fBlen\fR).
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIbio\fR\|(3)
.SH "HISTORY"
.IX Header "HISTORY"
\&\fIBUF_MEM_new()\fR, \fIBUF_MEM_free()\fR and \fIBUF_MEM_grow()\fR are available in all
versions of SSLeay and OpenSSL. \fIBUF_strdup()\fR was added in SSLeay 0.8.


third_part/linux_kernel_0.11/linux/fs/buffer.c
//  linux/fs/buffer.c
//  (C) 1991  Linus Torvalds

// 'buffer.c' implements the buffer-cache functions. Race-conditions have
// been avoided by NEVER letting a interrupt change a buffer (except for the
// data, of course), but instead letting the caller do it. NOTE! As interrupts
// can wake up a caller, some cli-sti sequences are needed to check for
// sleep-on-calls. These should be extremely quick, though (I hope).

// NOTE! There is one discordant note here: checking floppies for
// disk change. This is where it fits best, I think, as it should
// invalidate changed floppy-disk-caches.

#include <stdarg.h>
#include <linux/config.h>
#include <linux/sched.h>
#include <linux/kernel.h>
#include <asm/system.h>
#include <asm/io.h>

extern int end;
extern void put_super(int);
extern void invalidate_inodes(int);

struct buffer_head * start_buffer = (struct buffer_head *) &end;
struct buffer_head * hash_table[NR_HASH];
static struct buffer_head * free_list;
static struct task_struct * buffer_wait = NULL;
int NR_BUFFERS = 0;

static inline void wait_on_buffer(struct buffer_head * bh) {
	cli();
	while (bh->b_lock) sleep_on(&bh->b_wait);
	sti();
}

int sys_sync(void) {
	int i;
	struct buffer_head * bh;

	sync_inodes();		// write out inodes into buffers 
	bh = start_buffer;
	for (i = 0; i < NR_BUFFERS; i++,bh++) {
		wait_on_buffer(bh);
		if (bh->b_dirt) ll_rw_block(WRITE,bh);
	}
	return 0;
}

int sync_dev(int dev) {
	int i;
	struct buffer_head * bh;
	bh = start_buffer;
	for (i = 0; i < NR_BUFFERS; i++, bh++) {
		if (bh->b_dev != dev) continue;
		wait_on_buffer(bh);
		if (bh->b_dev == dev && bh->b_dirt) {
			ll_rw_block(WRITE,bh);
    }
	}
	sync_inodes();
	bh = start_buffer;
	for (i = 0; i < NR_BUFFERS; i++, bh++) {
		if (bh->b_dev != dev) continue;
		wait_on_buffer(bh);
		if (bh->b_dev == dev && bh->b_dirt) {
			ll_rw_block(WRITE,bh);
    }
	}
	return 0;
}

void inline invalidate_buffers(int dev) {
	int i;
	struct buffer_head * bh;

	bh = start_buffer;
	for (i = 0; i < NR_BUFFERS; i++, bh++) {
		if (bh->b_dev != dev) continue;
		wait_on_buffer(bh);
		if (bh->b_dev == dev) bh->b_uptodate = bh->b_dirt = 0;
	}
}

// This routine checks whether a floppy has been changed, and
// invalidates all buffer-cache-entries in that case. This
// is a relatively slow routine, so we have to try to minimize using
// it. Thus it is called only upon a 'mount' or 'open'. This
// is the best way of combining speed and utility, I think.
// People changing diskettes in the middle of an operation deserve
// to loose :-)
//
// NOTE! Although currently this is only for floppies, the idea is
// that any additional removable block-device will use this routine,
// and that mount/open needn't know that floppies/whatever are
// special.
void check_disk_change(int dev) {
	int i;
	if (MAJOR(dev) != 2) return;
	if (!floppy_change(dev & 0x03)) return;
	for (i = 0; i < NR_SUPER; i++) {
		if (super_block[i].s_dev == dev) {
			put_super(super_block[i].s_dev);
    }
  }
	invalidate_inodes(dev);
	invalidate_buffers(dev);
}

#define _hashfn(dev,block) (((unsigned)(dev^block))%NR_HASH)
#define hash(dev,block) hash_table[_hashfn(dev,block)]

static inline void remove_from_queues(struct buffer_head * bh) {
  // remove from hash-queue
	if (bh->b_next) bh->b_next->b_prev = bh->b_prev;
	if (bh->b_prev) bh->b_prev->b_next = bh->b_next;
	if (hash(bh->b_dev,bh->b_blocknr) == bh)
		hash(bh->b_dev,bh->b_blocknr) = bh->b_next;
  // remove from free list
	if (!(bh->b_prev_free) || !(bh->b_next_free)) {
		panic("Free block list corrupted");
  }
	bh->b_prev_free->b_next_free = bh->b_next_free;
	bh->b_next_free->b_prev_free = bh->b_prev_free;
	if (free_list == bh) {
		free_list = bh->b_next_free;
  }
}

static inline void insert_into_queues(struct buffer_head * bh) {
  // put at end of free list
	bh->b_next_free = free_list;
	bh->b_prev_free = free_list->b_prev_free;
	free_list->b_prev_free->b_next_free = bh;
	free_list->b_prev_free = bh;
  // put the buffer in new hash-queue if it has a device
	bh->b_prev = NULL;
	bh->b_next = NULL;
	if (!bh->b_dev) return;
	bh->b_next = hash(bh->b_dev,bh->b_blocknr);
	hash(bh->b_dev,bh->b_blocknr) = bh;
	bh->b_next->b_prev = bh;
}

static struct buffer_head * find_buffer(int dev, int block) {		
	struct buffer_head * tmp;
	for (tmp = hash(dev,  block); tmp != NULL; tmp = tmp->b_next) {
		if (tmp->b_dev==dev && tmp->b_blocknr==block) return tmp;
  }
	return NULL;
}

// Why like this, I hear you say... The reason is race-conditions.
// As we don't lock buffers (unless we are readint them, that is),
// something might happen to it while we sleep (ie a read-error
// will force it bad). This shouldn't really happen currently, but
// the code is ready.
struct buffer_head * get_hash_table(int dev, int block) {
	struct buffer_head * bh;
	for (; ;) {
		if (!(bh=find_buffer(dev, block))) return NULL;
		bh->b_count++;
		wait_on_buffer(bh);
		if (bh->b_dev == dev && bh->b_blocknr == block) {
			return bh;
    }
		bh->b_count--;
	}
}

// Ok, this is getblk, and it isn't very clear, again to hinder
// race-conditions. Most of the code is seldom used, (ie repeating),
// so it should be much more efficient than it looks.
// The algoritm is changed: hopefully better, and an elusive bug removed.
#define BADNESS(bh) (((bh)->b_dirt<<1)+(bh)->b_lock)
struct buffer_head * getblk(int dev,int block) {
	struct buffer_head * tmp, * bh;

repeat:
	if ((bh = get_hash_table(dev,block))) return bh;
	tmp = free_list;
	do {
		if (tmp->b_count) continue;
		if (!bh || BADNESS(tmp)<BADNESS(bh)) {
			bh = tmp;
			if (!BADNESS(tmp)) break;
		}
  // and repeat until we find something good
	} while ((tmp = tmp->b_next_free) != free_list);
	if (!bh) {
		sleep_on(&buffer_wait);
		goto repeat;
	}
	wait_on_buffer(bh);
	if (bh->b_count) goto repeat;
	while (bh->b_dirt) {
		sync_dev(bh->b_dev);
		wait_on_buffer(bh);
		if (bh->b_count) goto repeat;
	}
  // NOTE!! While we slept waiting for this block, somebody else might
  // already have added "this" block to the cache. check it
	if (find_buffer(dev,block)) goto repeat;
  // OK, FINALLY we know that this buffer is the only one of it's kind,
  // and that it's unused (b_count=0), unlocked (b_lock=0), and clean
	bh->b_count = 1;
	bh->b_dirt = 0;
	bh->b_uptodate = 0;
	remove_from_queues(bh);
	bh->b_dev = dev;
	bh->b_blocknr = block;
	insert_into_queues(bh);
	return bh;
}

void brelse(struct buffer_head * buf) {
	if (!buf) return;
	wait_on_buffer(buf);
	if (!(buf->b_count--)) {
		panic("Trying to free free buffer");
  }
	wake_up(&buffer_wait);
}

// bread() reads a specified block and returns the buffer that contains
// it. It returns NULL if the block was unreadable.
struct buffer_head * bread(int dev, int block) {
	struct buffer_head * bh;
	if (!(bh=getblk(dev, block))) {
		panic("bread: getblk returned NULL\n");
  }
	if (bh->b_uptodate) return bh;
	ll_rw_block(READ,bh);
	wait_on_buffer(bh);
	if (bh->b_uptodate) return bh;
	brelse(bh);
	return NULL;
}

#define COPYBLK(from, to) \
__asm__("cld\n\t" \
	"rep\n\t" \
	"movsl\n\t" \
	::"c" (BLOCK_SIZE/4),"S" (from),"D" (to) \
	)

// bread_page reads four buffers into memory at the desired address. It's
// a function of its own, as there is some speed to be got by reading them
// all at the same time, not waiting for one to be read, and then another
// etc.
void bread_page(unsigned long address,int dev,int b[4]) {
	struct buffer_head * bh[4];
	int i;
	for (i = 0; i < 4; i++) {
		if (b[i]) {
			if ((bh[i] = getblk(dev,b[i]))) {
				if (!bh[i]->b_uptodate) ll_rw_block(READ,bh[i]);
      }
		} else bh[i] = NULL;
  }
	for (i = 0; i < 4; i++, address += BLOCK_SIZE)
		if (bh[i]) {
			wait_on_buffer(bh[i]);
			if (bh[i]->b_uptodate) {
				COPYBLK((unsigned long) bh[i]->b_data,address);
      }
			brelse(bh[i]);
		}
}
// Ok, breada can be used as bread, but additionally to mark other
// blocks for reading as well. End the argument list with a negative
// number.
struct buffer_head * breada(int dev,int first, ...) {
	va_list args;
	struct buffer_head * bh, *tmp;

	va_start(args,first);
	if (!(bh=getblk(dev,first))) {
		panic("bread: getblk returned NULL\n");
  }
	if (!bh->b_uptodate) ll_rw_block(READ,bh);
	while ((first = va_arg(args, int)) >= 0) {
		tmp = getblk(dev,first);
		if (tmp) {
			if (!tmp->b_uptodate) {
				ll_rw_block(READA,bh);
      }
			tmp->b_count--;
		}
	}
	va_end(args);
	wait_on_buffer(bh);
	if (bh->b_uptodate) return bh;
	brelse(bh);
	return (NULL);
}

void buffer_init(long buffer_end) {
	struct buffer_head * h = start_buffer;
	void * b;
	int i;

	if (buffer_end == 1<<20) b = (void *) (640*1024);
	else b = (void *) buffer_end;
	while ((b -= BLOCK_SIZE) >= ((void *) (h+1))) {
		h->b_dev = 0;
		h->b_dirt = 0;
		h->b_count = 0;
		h->b_lock = 0;
		h->b_uptodate = 0;
		h->b_wait = NULL;
		h->b_next = NULL;
		h->b_prev = NULL;
		h->b_data = (char *) b;
		h->b_prev_free = h-1;
		h->b_next_free = h+1;
		h++;
		NR_BUFFERS++;
		if (b == (void *) 0x100000) {
			b = (void *) 0xA0000;
    }
	}
	h--;
	free_list = start_buffer;
	free_list->b_prev_free = h;
	h->b_next_free = free_list;
	for (i = 0; i < NR_HASH; i++) {
		hash_table[i] = NULL;
  }
}	
